ext/standard/proc_open.c

```diff
@@ -465,7 +467,269 @@ static char *create_win_command_from_args(HashTable *args) {
+		}
+	}
+}
+
+static void init_startup_info(STARTUPINFOW *si, struct php_proc_open_descriptor_item *descriptors, int ndesc)
+{
+	memset(si, 0, sizeof(STARTUPINFOW));
+	si->cb = sizeof(STARTUPINFOW);
+	si->dwFlags = STARTF_USESTDHANDLES;
+
+	si->hStdInput  = GetStdHandle(STD_INPUT_HANDLE);
+	si->hStdOutput = GetStdHandle(STD_OUTPUT_HANDLE);
+	si->hStdError  = GetStdHandle(STD_ERROR_HANDLE);
+
+	/* redirect stdin/stdout/stderr if requested */
+	for (int i = 0; i < ndesc; i++) {
+		switch(descriptors[i].index) {
+			case 0:
+				si->hStdInput = descriptors[i].childend;
+				break;
+			case 1:
+				si->hStdOutput = descriptors[i].childend;
+				break;
+			case 2:
+				si->hStdError = descriptors[i].childend;
+				break;
+		}
+	}
+}
+
+static void init_process_info(PROCESS_INFORMATION *pi)
+{
+	memset(&pi, 0, sizeof(pi));
+}
+
+static int convert_command_to_use_shell(wchar_t **cmdw, size_t cmdw_len)
+{
+	size_t len = sizeof(COMSPEC_NT) + sizeof(" /s /c ") + cmdw_len + 3;
+	wchar_t *cmdw_shell = (wchar_t *)malloc(len * sizeof(wchar_t));
+
+	if (cmdw_shell == NULL) {
+		php_error_docref(NULL, E_WARNING, "Command conversion failed");
+		return FAILURE;
+	}
+
+	if (_snwprintf(cmdw_shell, len, L"%hs /s /c \"%s\"", COMSPEC_NT, *cmdw) == -1) {
+		free(cmdw_shell);
+		php_error_docref(NULL, E_WARNING, "Command conversion failed");
+		return FAILURE;
+	}
+
+	free(*cmdw);
+	*cmdw = cmdw_shell;
+
+	return SUCCESS;
+}
+#endif
+
+static int get_command_from_array(zval *array, char ***argv, char **command, int num_elems, int is_persistent)
+{
+	zval *arg_zv;
+	int i = 0;
+
+	*argv = safe_emalloc(sizeof(char *), num_elems + 1, 0);
+
+	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(array), arg_zv) {
+		zend_string *arg_str = get_valid_arg_string(arg_zv, i + 1);
+		if (!arg_str) {
+			/* terminate argv with NULL so exit_fail code knows how many entries to free */
+			(*argv)[i] = NULL;
+			return FAILURE;
+		}
+
+		if (i == 0) {
+			*command = pestrdup(ZSTR_VAL(arg_str), is_persistent);
+		}
+
+		(*argv)[i++] = estrdup(ZSTR_VAL(arg_str));
+		zend_string_release(arg_str);
+	} ZEND_HASH_FOREACH_END();
+
+	(*argv)[i] = NULL;
+
+	/* As the array is non-empty, we should have found a command. */
+	ZEND_ASSERT(command);
+
+	return SUCCESS;
+}
+
+static struct php_proc_open_descriptor_item* alloc_descriptor_array(zval *descriptorspec)
+{
+	int ndescriptors = zend_hash_num_elements(Z_ARRVAL_P(descriptorspec));
+	struct php_proc_open_descriptor_item *descriptors = safe_emalloc(sizeof(struct php_proc_open_descriptor_item), ndescriptors, 0);
+	memset(descriptors, 0, sizeof(struct php_proc_open_descriptor_item) * ndescriptors);
+	return descriptors;
+}
+
+static int get_string_parameter(zval **dest, zval *array, int index, char *param_name)
+{
+	if ((*dest = zend_hash_index_find(Z_ARRVAL_P(array), index)) == NULL) {
+		zend_value_error("Missing %s", param_name);
+		return FAILURE;
+	}
+	if (!try_convert_to_string(*dest)) {
+		zend_value_error("Cannot convert %s to string", param_name);
```

Comment:
This shouldn't be here, try_convert_to_string already throws.

We should also convert this code to use zval_try_get_string and return a zend_string*.

Reply:
Will do.